packages/gamut-labs/src/Flyout/index.tsx

```diff
@@ -0,0 +1,157 @@
+import { BodyPortal, FocusTrap, IconButton } from '@codecademy/gamut';
+import { MiniDeleteIcon } from '@codecademy/gamut-icons';
+import { system, variant } from '@codecademy/gamut-styles';
+import { StyleProps, variance } from '@codecademy/variance';
+import styled from '@emotion/styled';
+import { AnimatePresence, motion } from 'framer-motion';
+import React, { useCallback, useEffect, useState } from 'react';
+
+export const transitionDuration = 0.35;
+
+const flyoutStyles = variance.compose(
+  system.background,
+  system.border,
+  system.color,
+  system.flex,
+  system.grid,
+  system.layout,
+  system.positioning
+);
+
+const flyoutOpenVariants = variant({
+  defaultVariant: 'left',
+  prop: 'openFrom',
+  variants: {
+    left: {
+      right: 'auto',
+      left: 0,
+    },
+    right: {
+      right: 0,
+      left: 'auto',
+    },
+  },
+});
+
+export type FlyoutStyles = StyleProps<typeof flyoutStyles> &
+  StyleProps<typeof flyoutOpenVariants>;
+export interface FlyoutStyleProps extends FlyoutStyles {}
+
+export const DrawerBase = styled(motion.div)<FlyoutStyleProps>(
+  flyoutStyles,
+  flyoutOpenVariants
+);
+
+type FlyoutProps = FlyoutStyleProps & {
+  /**
+   * if the drawer should be open or closed
+   */
+  initialExpanded?: boolean;
+  /**
+   * data-testid for the components
+   */
+  testId?: string;
+  /**
+   * width of the open drawer in rem
+   */
+  openWidth?: number;
+  /**
+   * toggles the Flyout
+   */
+  renderButton: (onClick: () => void) => React.ReactNode;
+  /**
+   * Whether clicking on the screen outside of the container should close the Flyout
+   */
+  clickOutsideDoesNotClose?: boolean;
+  /**
+   * Whether clicking the escape key should close the Flyout
+   */
+  escapeDoesNotClose?: boolean;
+
+  /**
+   * A means of the parent method to get a reference to the closeFlyout function
+   */
+  closeFlyoutRef?: React.MutableRefObject<Function>;
+};
+
+export const Flyout: React.FC<FlyoutProps> = ({
+  children,
+  renderButton,
+  initialExpanded,
+  openFrom = 'left',
+  openWidth = 30,
+  testId,
+  clickOutsideDoesNotClose,
+  escapeDoesNotClose,
+  closeFlyoutRef,
+  ...styleProps
+}) => {
+  const initialX = openFrom === 'left' ? -1000 : 1000;
+
+  const [isExpanded, setIsExpanded] = useState(!!initialExpanded);
+  const toggleExpanded = useCallback(
+    () => setIsExpanded((isExpanded) => !isExpanded),
+    []
+  );
+  const closeFlyout = useCallback(() => {
+    setIsExpanded(false);
+  }, []);
+
+  useEffect(() => {
+    // Passes the function up to any interested parent component
+    if (closeFlyoutRef) {
+      closeFlyoutRef.current = closeFlyout;
+      return () => {
+        closeFlyoutRef.current = () => {};
```

Comment:
Nit: maybe make this `(() => void) | undefined` and make the parent have to `?.()` call it, to be more explicit about it not existing maybe?

```suggestion
        closeFlyoutRef.current = undefined;
```

Reply:
Josh did you mean like this?
https://github.com/Codecademy/client-modules/pull/1740/commits/45d088c0af790cbed9d3892ea1532206cdfd1906

Reply:
CC: @MattSchiller

Reply:
Reverted the changes in https://github.com/Codecademy/client-modules/commit/45d088c0af790cbed9d3892ea1532206cdfd1906 to 

https://github.com/Codecademy/client-modules/pull/1740/commits/130b84631e3202547953a21fcdd3a2df66be52ad.

@MattSchiller put it as so:

> before, we could at least ensure that it was always a function
> but now it might be undefined
> so callers need to protect themselves, and without type safety there, too